### PR TITLE
Update swift-demangle to interactively demangle

### PR DESCRIPTION
**Pull request by keith**
#### What's in this pull request?

This updates `swift-demangle`'s interactive stdin mode to immediately output demangled strings as opposed to waiting until ctrl-d is pressed.
#### Resolved bug number: ([SR-2410](https://bugs.swift.org/browse/SR-2410))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

This changes the swift-demangle interactive STDIN mode to immediately
print the demangled symbol on return as opposed to printing them all at
the end.

### CODE REVIEW

File: tools/swift-demangle/swift-demangle.cpp [hunk 1]
```diff
@@ -86,22 +102,45 @@ static void demangle(llvm::raw_ostream &os, llvm::StringRef name,
...
 
-static llvm::StringRef substrAfter(llvm::StringRef whole,
-                                   llvm::StringRef part) {
-  return whole.substr((part.data() - whole.data()) + part.size());
+  while (true) {
+    llvm::SmallString<ChunkSize> Buffer;
+    ssize_t ReadBytes;
+
+    Buffer.reserve(Buffer.size() + ChunkSize);
+    ReadBytes = read(STDIN_FILENO, Buffer.end(), ChunkSize);
```
Comment (jrose-apple) [inline tools/swift-demangle/swift-demangle.cpp]: I'm concerned about breaking a symbol across chunk boundaries. Can we do something line-oriented instead? (I don't think we care about running this tool across binary data.)

**Comment by jrose-apple** [comment]
Thanks, Keith! Hm, is there any way to write a test for this?

**Comment by keith** [comment]
Thanks for the feedback Jordan! I've pushed a `getline` approach and some tests, let me know what you think!

**Comment by jrose-apple** [comment]
This implementation looks good. I think we already have tests for stdin-demangling in test/Demangle/; what I really wanted to know is if there's a way to test that it's streaming now. I can't think of a good way without building a whole wrapper script, though—which would be nice, but I wouldn't block the patch over it. I think it's okay to drop this extra test file and just move ahead.

**Comment by keith** [comment]
Ah sorry about that. I looked in that directory but assumed tests would be named after the CLI. I'll remove those here!

**Comment by keith** [comment]
Updated

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

**Comment by shahmishal** [comment]
@swift-ci Please smoke test OS X

**Comment by jrose-apple** [comment]
Merged. Thanks, Keith!

**Comment by keith** [comment]
Thanks! Should we close https://bugs.swift.org/browse/SR-2410 ?

**Comment by CodaFi** [comment]
Yes, please.

**Pull request merged**